### PR TITLE
[FLINK-29035][table-planner] Fix bug of ExpressionReducer does not work with jar resources

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

Fix bug of ExpressionReducer does not work with jar resources

## Brief change log

  - *Fix bug of ExpressionReducer does not work with jar resources*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests in FunctionITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by lsyldliu** [comment]
cc @twalthr

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java [hunk 1]
```diff
@@ -159,6 +159,22 @@ public void testUsingAddJar() throws Exception {
                 "drop function lowerUdf");
     }
 
+    @Test
+    public void testExpressionReducerByUsingJar() {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java:163]: We should avoid code deduplication in the code base. Most paths around functions are identical for both batch and streaming. So I don't think we need dedicated tests for `batch/sql/FunctionITCase`. What do you think about removing `batch/sql/FunctionITCase` entirely (my recommendation) or at have just one basic feature test in there?

**Review by twalthr — CHANGES_REQUESTED** [review]
The change looks good to me. I just had a minor comment regarding testing.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java (hunk 1 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/batch/sql/FunctionITCase.java:163]: Yes, I agree with you that we should avoid code deduplication, but I prefer the latter, we should have one test to prove it works in batch mode. BTW, maybe there exists some batch exclusive test in the future.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**